x-pack/auditbeat/module/system/login/utmp.go

```diff
@@ -0,0 +1,571 @@
+
+// +build linux,cgo
+
+package login
+
+// #include <utmp.h>
+// #include <stdlib.h>
+import "C"
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/gob"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"os/user"
+	"path/filepath"
+	"reflect"
+	"sort"
+	"strconv"
+	"syscall"
+	"time"
+	"unsafe"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/auditbeat/datastore"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+const (
+	bucketKeyFileRecords   = "file_records"
+	bucketKeyLoginSessions = "login_sessions"
+)
+
+// Inode represents a file's inode on Linux.
+type Inode uint64
+
+// FileRecord represents a UTMP file at a point in time.
+type FileRecord struct {
+	Inode    Inode
+	Size     int64
+	LastUtmp Utmp
+}
+
+// Utmp contains data from the C utmp struct.
+type Utmp struct {
+	UtType   int
+	UtPid    int
+	UtLine   string
+	UtUser   string
+	UtHost   string
+	UtTv     time.Time
+	UtAddrV6 [4]uint32
+}
+
+func newUtmp(utmpC *C.struct_utmp) Utmp {
+	// See utmp(5) for the utmp struct fields.
+	return Utmp{
+		UtType:   int(utmpC.ut_type),
+		UtPid:    int(utmpC.ut_pid),
+		UtLine:   C.GoString(&utmpC.ut_line[0]),
+		UtUser:   C.GoString(&utmpC.ut_user[0]),
+		UtHost:   C.GoString(&utmpC.ut_host[0]),
+		UtTv:     time.Unix(int64(utmpC.ut_tv.tv_sec), int64(utmpC.ut_tv.tv_usec)*1000),
+		UtAddrV6: [4]uint32{uint32(utmpC.ut_addr_v6[0]), uint32(utmpC.ut_addr_v6[1]), uint32(utmpC.ut_addr_v6[2]), uint32(utmpC.ut_addr_v6[3])},
+	}
+}
+
+// UtmpFileReader can read a UTMP formatted file (usually /var/log/wtmp).
+type UtmpFileReader struct {
+	log           *logp.Logger
+	bucket        datastore.Bucket
+	filePattern   string
+	fileRecords   map[Inode]FileRecord
+	loginSessions map[string]LoginRecord
+}
+
+// NewUtmpFileReader creates and initializes a new UTMP file reader.
+func NewUtmpFileReader(log *logp.Logger, bucket datastore.Bucket, filePattern string) (*UtmpFileReader, error) {
+	r := &UtmpFileReader{
+		log:           log,
+		bucket:        bucket,
+		filePattern:   filePattern,
+		fileRecords:   make(map[Inode]FileRecord),
+		loginSessions: make(map[string]LoginRecord),
+	}
+
+	// Load state (file records, tty mapping) from disk
+	err := r.restoreStateFromDisk()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to restore state from disk")
+	}
+
+	return r, nil
+}
+
+// Close performs any cleanup tasks when the UTMP reader is done.
+func (r *UtmpFileReader) Close() error {
+	err := r.bucket.Close()
+	if err != nil {
+		return errors.Wrap(err, "error closing bucket")
+	}
+
+	return nil
+}
+
+// ReadNew returns any new UTMP entries in any files matching the configured pattern.
+func (r *UtmpFileReader) ReadNew() ([]LoginRecord, error) {
+	fileInfos, err := r.fileInfos()
+	defer r.deleteOldRecords(fileInfos)
+
+	var loginRecords []LoginRecord
+	for path, fileInfo := range fileInfos {
+		inode := Inode(fileInfo.Sys().(*syscall.Stat_t).Ino)
+
+		fileRecord, isKnownFile := r.fileRecords[inode]
+		var oldSize int64
```

Comment:
You can skip the isKnownFile check here. The value of Size will be 0 if it's not found in the map.
```suggestion
		oldSize := fileRecord.Size
```